sql/types/json_test.go

```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
+		// +--------------------------------------------------+
+		// 1 row in set (0.00 sec)
+		desc:      "no update for element underflow",
+		doc:       `[1, 2, 3]`,
+		path:      "$[last-23]",
+		value:     `42`,
+		resultVal: `[1, 2, 3]`,
+		changed:   false,
+	},
+	{
+		desc:      "no update for empty array",
+		doc:       `[]`,
+		path:      "$[0]",
+		value:     `42`,
+		resultVal: `[]`,
+		changed:   false,
+	},
+	{
+		desc:      "treating object as an array replaces for index 0",
+		doc:       `{"a":1}`,
+		path:      "$[0]",
+		value:     `42`,
+		resultVal: `42`,
+		changed:   true,
+	},
+	{
+		// mysql> select JSON_REPLACE(JSON_OBJECT("a",1),'$[last]', 42);
+		// +------------------------------------------------+
+		// | JSON_REPLACE(JSON_OBJECT("a",1),'$[last]', 42) |
+		// +------------------------------------------------+
+		// | 42                                             |
+		// +------------------------------------------------+
+		desc:      "treating object as an array replaces for index last",
+		doc:       `{"a":1}`,
+		path:      "$[last]",
+		value:     `42`,
+		resultVal: `42`,
+		changed:   true,
+	},
+	{
+		desc:      "no op when treating object as an array with underflow",
+		doc:       `{"a":1}`,
+		path:      "$[last-23]",
+		value:     `42`,
+		resultVal: `{"a": 1}`,
+		changed:   false,
+	},
+	{
+		desc:      "no op when treating object as an array with overflow",
+		doc:       `{"a":1}`,
+		path:      "$[51]",
+		value:     `42`,
+		resultVal: `{"a": 1}`,
+		changed:   false,
+	},
+	{
+		desc:      "no update for scalar will treated as an array for out of bounds",
+		doc:       `17`,
+		path:      "$[51]",
+		value:     `42`,
+		resultVal: `17`,
+		changed:   false,
+	},
+	{
+		desc:      "scalar will be overwritten for index 0",
+		doc:       `17`,
+		path:      "$[0]",
+		value:     `42`,
+		resultVal: `42`,
+		changed:   true,
+	},
+	{
+		desc:      "no update for scalar when used as an array with underflow",
+		doc:       `17`,
+		path:      "$[last-23]",
+		value:     `42`,
+		resultVal: `17`,
+		changed:   false,
+	},
+	{
+		desc:      "Object field updated",
+		doc:       `{"a": 1}`,
+		path:      "$.a",
+		value:     `42`,
+		resultVal: `{"a": 42}`,
+		changed:   true,
+	},
+	{
+		desc:      "Object field not inserted",
+		doc:       `{"a": 1}`,
+		path:      "$.b",
+		value:     `42`,
+		resultVal: `{"a": 1}`,
+		changed:   false,
+	},
+}
+
+func TestJsonReplace(t *testing.T) {
+	for _, test := range JsonReplaceTests {
+		t.Run("JSON replace: "+test.desc, func(t *testing.T) {
+			doc := MustJSON(test.doc)
+			val := MustJSON(test.value)
+			res, changed, err := doc.Replace(sql.NewEmptyContext(), test.path, val)
+			require.NoError(t, err)
+			assert.Equal(t, MustJSON(test.resultVal), res)
+			assert.Equal(t, test.changed, changed)
+		})
+	}
+}
+
+type parseErrTest struct {
+	desc         string
+	path         string
+	expectErrStr string
+}
+
+var JsonPathParseErrTests = []parseErrTest{
```

Comment:
Should we add cases for using `*` or `**` wildcards? I noticed several json functions mentioned those are not allowed in the paths.

Reply:
I'm going to handle this in the next PR.